File: labs/caminhao/src/cpp/solver.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long lli;

lli gasolina[100000 + 10]; // Armazena a quantidade de gasolina disponível em cada bomba
lli distancia[100000 + 10]; // Armazena a distância até a próxima bomba
queue<int> filaDeBombas; // Fila para armazenar os índices das bombas consideradas
long unsigned int numBombas; // Número total de bombas

// Macro para calcular o saldo de gasolina em uma bomba específica
#define saldo(bomba) (gasolina[bomba] - distancia[bomba])

int main(void) {
  long unsigned int i;
  cin >> numBombas;

  // Lendo os valores de gasolina e distância para cada bomba
  for (i = 0; i < numBombas; i++)
    cin >> gasolina[i] >> distancia[i];

  // Garantindo que a fila esteja vazia
  while (!filaDeBombas.empty())
    filaDeBombas.pop();

  filaDeBombas.push(0); // Começamos testando a partir da bomba 0

  lli saldoAcumulado = 0; // Mantém o saldo de gasolina enquanto percorremos as bombas
  int ultimoIndice = 0; // Última bomba considerada na tentativa de completar o circuito

  while (filaDeBombas.size() < numBombas) {
    ultimoIndice = filaDeBombas.back(); // Obtém o índice da última bomba adicionada à fila
    saldoAcumulado += saldo(ultimoIndice); // Atualiza o saldo de gasolina

    // Se o saldo for negativo, significa que não conseguimos avançar, então removemos bombas
    while (saldoAcumulado < 0 && !filaDeBombas.empty())
      filaDeBombas.pop();

    // Se a fila ficou vazia, reiniciamos o saldo
    if (filaDeBombas.empty())
      saldoAcumulado = 0;

    // Adicionamos a próxima bomba à fila
    filaDeBombas.push((ultimoIndice + 1) % numBombas);
  }

  // O primeiro elemento da fila é o ponto de partida ideal
  printf("%d\n", filaDeBombas.front());

  return 0;
}

File: labs/caminhao/src/go/solver.go
package main

import (
	"container/list"
	"fmt"
)

// Função para encontrar o ponto de partida ideal
func findStartingPump(gasoline []int, distance []int, n int) int {
	queue := list.New() // Lista encadeada para simular a fila
	queue.PushBack(0)   // Começamos testando pela bomba 0

	saldoAcumulado := 0 // Mantém o saldo de gasolina acumulado
	ultimoIndice := 0   // Última bomba considerada na tentativa de completar o circuito

	for queue.Len() < n {
		ultimoIndice = queue.Back().Value.(int) // Obtém o último índice da fila
		saldoAcumulado += gasoline[ultimoIndice] - distance[ultimoIndice]

		// Se o saldo for negativo, removemos bombas da frente da fila até o saldo ser positivo
		for saldoAcumulado < 0 && queue.Len() > 0 {
			queue.Remove(queue.Front())
		}

		// Se a fila esvaziou, reiniciamos o saldo
		if queue.Len() == 0 {
			saldoAcumulado = 0
		}

		// Adicionamos a próxima bomba à fila
		queue.PushBack((ultimoIndice + 1) % n)
	}

	// O primeiro elemento da fila é o ponto de partida ideal
	return queue.Front().Value.(int)
}

func main() {
	var n int
	fmt.Scan(&n)

	gasoline := make([]int, n)
	distance := make([]int, n)

	for i := 0; i < n; i++ {
		fmt.Scan(&gasoline[i], &distance[i])
	}

	fmt.Println(findStartingPump(gasoline, distance, n))
}
